Support customization of fields on derive

greyblake opened the pull request:
This PR addresses https://github.com/rust-fuzz/arbitrary/issues/33 and enables the following syntax for derive:

```rust
#[derive(Arbitrary)]
pub struct Rgb {
    // set `r` to Default::default()
    #[arbitrary(default)]
    pub r: u8,

    // set `g` to 255
    #[arbitrary(value = "255")]
    pub g: u8,

    // generate `b` with a custom function
    #[arbitrary(with = "arbitrary_b")]
    pub b: u8,
}

fn arbitrary_b(u: &mut Unstructured) -> arbitrary::Result<u8> {
    u.int_in_range(64..=128)
}
```

## Note
In case of custom function the lower bound for `hint_size` is defined through `core::mem::size_of::<T>()` which is correct for the types allocated on stack, but not for the heap allocated types.

fitzgen commented on README.md:
```diff
@@ -61,6 +61,31 @@ pub struct Rgb {
+    pub r: u8,
+
+    // set `g` to 255
+    #[arbitrary(value = "255")]
```
I think we should be able to allow the syntax

```rust
#[arbitrary(value = 255)]
```

here. That is, just pass through any tokens on the RHS, so users could even do function calls and stuff like `value = 1 + foo()` or whatever. This should also make the `parse_assigned_value` function unnecessary.

Manishearth replied:
+1

greyblake replied:
My inspiration was taken from [serde](https://serde.rs/field-attrs.html)
```rust
#[serde(deserialize_with = "path")] 
```
I think the reasoning for this is if the values are double quoted it's much easier to parse second and further optional attributes if necessary. 
Anyway, I will change it.

greyblake replied:
Addressed in https://github.com/rust-fuzz/arbitrary/pull/129/commits/f73d5f2333f4004ccdd5250c198044e5102d61f9

fitzgen commented on tests/derive.rs:
```diff
@@ -231,3 +231,42 @@ fn recursive_and_empty_input() {
 }
+
+#[test]
+fn test_field_attributes() {
```
Can we also add some doc tests that use `compile_fail` to check that we return compiler errors when

* `#[arbitrary]` is used multiple times on a single field
* `#[arbitrary(unknown_attr)]` and `#[arbitrary(unknown_attr = unknown_val)]` are used
* `#[arbitrary(value)]` and `#[arbitrary(with)]` are used without RHS values

Should be able to just do something like the following in this file:

```rust
/// Can only use `#[arbitrary]` once per field:
///
/// ```compile_fail
/// use arbitrary::*;
/// #[derive(Arbitrary)]
/// struct Foo {
///     #[arbitrary(with = foo)]
///     #[arbitrary(with = bar)]
///     field: u32,
/// }
/// fn foo(_: &mut Unstructured) -> u32 { todo!() }
/// fn bar(_: &mut Unstructured) -> u32 { todo!() }
/// ```
///
/// Etc...
pub struct CompileFailTests;
```

greyblake replied:
Addressed in https://github.com/rust-fuzz/arbitrary/pull/129/commits/8913ec0d40af472deec4e4ab01eb63ee69d73c80

I was puzzled where to put such doc tests (I don't think users really want to see them in the docs).
Eventually I look into `serde` again, they're using [trybuild](https://github.com/dtolnay/trybuild) to test compile failures. 
So I've decided to introduce `trybuild` to this project too, I think it's a nice tool for this purpose.

fitzgen replied:
Was doing doc tests on a `pub struct CompileFailTests` in `tests/derive.rs` not getting picked up by rustdoc?

If not, then we could maybe avoid a new dependency by doing

```rust
// src/lib.rs

#[cfg(all(test, feature = "derive"))]
/// ...
pub struct CompileFailTests {}
```

So that it never shows up in the docs and only is built when we are running tests (with the derive enabled).

greyblake replied:
@fitzgen I just tried doc tests within `derive/src/lib.rs` are not picked up.

> If not, then we could maybe avoid a new dependency by doing

It's `dev-dependency`, shouldn't it be fine? I definitely like `trybuild`, it generates really helpful error messages and allows to structure tests in more easy to understand / maintainable way.

For example, this is what a reported failure looks like:
![image](https://user-images.githubusercontent.com/113512/197013993-cdd9c7f1-9c8e-4c25-a485-9a3613c27062.png)

This way we actually test, the compiler fails in the way we expect it to fail.  Doc tests may give false positives if something in the code gets mistyped.

I would see the cost of using `trybuild` is rather potential failures  if `rustc` changes the way it prints errors.
But still I think pros of `trybuild` outweights its cons.

Please let me know, if you still insist on using doc tests.

fitzgen replied:
I think if the test is in `src/lib.rs` rather than `derive/src/lib.rs` it should work? I'd really want to avoid failing tests on nightly but not stable unless as an absolute last resort.

fitzgen commented on README.md:
```diff
@@ -61,6 +61,31 @@ pub struct Rgb {
+    pub g: u8,
+
+    // generate `b` with a custom function
+    #[arbitrary(with = "arbitrary_b")]
```
Similarly, I think this should be an ident or path rather than a string, and we can implement this by just passing the RHS tokens through rather than parsing them. For example:

```rust
#[arbitrary(with = super::foo::bar::my_func)]
```

greyblake replied:
Addressed in https://github.com/rust-fuzz/arbitrary/pull/129/commits/f73d5f2333f4004ccdd5250c198044e5102d61f9

fitzgen commented on README.md:
```diff
@@ -61,6 +61,31 @@ pub struct Rgb {
+    #[arbitrary(value = "255")]
+    pub g: u8,
+
+    // generate `b` with a custom function
```
And I think we should document the type of the expected function, something like

```suggestion
    // Generate `b` with a custom function of type
    //
    //     fn(&mut Unstructured) -> arbitrary::Result<T>
    //
    // where `T` is the field's type.
```

greyblake replied:
Addressed in https://github.com/rust-fuzz/arbitrary/pull/129/commits/386de854122cbfab158bd99ffd7b7f4c46c8cfe9

fitzgen commented on README.md:
````diff
@@ -61,6 +61,31 @@ pub struct Rgb {
 }
 ```
 
+### Customizing single fields
````
This is a subsection of the `Automatically Deriving Arbitrary` section so it should have one more `#`:

```suggestion
#### Customizing single fields
```

greyblake replied:
Addressed in https://github.com/rust-fuzz/arbitrary/pull/129/commits/386de854122cbfab158bd99ffd7b7f4c46c8cfe9

fitzgen commented on derive/src/field_attributes.rs:
```diff
@@ -0,0 +1,110 @@
+            let value = parse_assigned_value(tokens_iter);
+            FieldConstructor::Value(value)
+        }
+        _ => panic!("Unknown options for {ARBITRARY_ATTRIBUTE_NAME}: {token}"),
```
Rather than panicking, can we return actual `Result`s and then in the main derive entry point and then either use https://docs.rs/syn/latest/syn/parse/struct.Error.html#method.to_compile_error or define our own `impl ToTokens for MyError` equivalent that expands into

```rust
quote! { compile_error!(#the_error_string) }
```

I think there isn't any reason we can't use `syn::Error` though.

This will provide better diagnostics for users.

greyblake replied:
Thanks. I was not aware of `syn::Error`, a cool thing, I will added it to my toolbox.

Addressed within the following commits:
- https://github.com/rust-fuzz/arbitrary/pull/129/commits/f3225dddfead7034441829846394b6b30604f4c8 - refactor functions to return `syn::Result<TokenStream>`
- https://github.com/rust-fuzz/arbitrary/pull/129/commits/fb58d7a992bbcb42ebaedb0f23b2de435d41809e, https://github.com/rust-fuzz/arbitrary/pull/129/commits/8cbad14fd9c672413f15d3d53d1b2072cf72fbc9, https://github.com/rust-fuzz/arbitrary/pull/129/commits/479da322a5b65917bb933dc43b03a83ecdec0c03 - get rid of `panic!()`, return `syn::Error`.

This indeed resulted into better error messages :heart: . You can see the difference in `.stderr` files.

fitzgen commented on tests/derive.rs:
```diff
@@ -231,3 +231,42 @@ fn recursive_and_empty_input() {
+
+    #[derive(Debug, Arbitrary)]
+    struct Parcel {
+        #[arbitrary(with = "arbitrary_weight")]
```
Can we also add a test where users define a closure for the custom arbitrary function?

For example:

```rust
#[arbitrary(with = |u| u.int_in_range(0..=10))]
```

I could see the generated code needing to wrap parens or put the function in a temporary variable to avoid compile errors in the generated code with this kind of macro input.

greyblake replied:
Addressed in https://github.com/rust-fuzz/arbitrary/pull/129/commits/727ab88363108e041243375c5ed2e31389a78d32

Note, the compiler require a type hint for `u`.
So 
```rust
#[arbitrary(with = |u| u.int_in_range(0..=10))]
```
does not work,
but 
```rust
#[arbitrary(with = |u: &mut Unstructured| u.int_in_range(0..=10))]
```
does.

There is a chance to make it work without specifying the type for `u`,  but it would require some dancing around to help the compiler to derive  the type (e.g. wrapping a closure into a helper function, etc.). I think it will be too much effort and mess (in the generated code) for a very little value.

fitzgen reviewed (commented):
Thanks! A few suggestions below before we merge this in.

greyblake commented:
The remaining part is using `syn::Result` instead of panic. I will finish it today later.

greyblake commented:
@fitzgen , @Manishearth I've addressed your comments guys. Thanks to you even learned a few new things today.

The easiest way to review the changes will be to review them commit by commit.

If you merge this PR, I would very very appreciate releasing a new version of the crate, so I can benefit from the changes in my main project.
Thanks.

fitzgen commented on tests/ui/multiple_arbitrary.stderr:
```diff
@@ -1,7 +1,5 @@
-7 | |     x: i32,
-  | |__________^
+5 |     #[arbitrary(value = 2)]
+  |     ^
```
Does this mean that the tests won't pass on both nightly and stable at the same time right now?

greyblake replied:
That's correct. I got one tests broken on CI, because I was running `nightly` locally, but CI was running `stable`.

OK, i'll change it to doc tests then.

fitzgen reviewed (commented):
Thanks! Looking great.

So `trybuild` is kinda nice because it lets you check the error messages, but if we can't have the tests passing on both stable and nightly at the same time, then I'd prefer exploring the doc tests work around instead.

greyblake commented:
@fitzgen Ok, here are the doc tests: https://github.com/rust-fuzz/arbitrary/pull/129/commits/755b8e0f5856acb4d6eb9b9b8d4fabcb16564159

fitzgen reviewed (approved):
Thanks!!

fitzgen commented:
Yeah I can cut a release in a little bit

fitzgen commented:
published

fitzgen commented:
Thanks again!

greyblake commented:
@fitzgen Thank you!
